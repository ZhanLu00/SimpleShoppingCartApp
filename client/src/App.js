import React from 'react';
import './App.css';
import Navbar from './React-Components/Navbar';
import AppContainer from './React-Components/AppContainer';
import { getSubtotal, checkout } from './Actions/action';

class Item{
	constructor(name, quantity, price, id){
		this.name = name;
		this.quantity = quantity;
		this.price = price;
		this.id = id;
	}

	decreaseId(){
		this.id-=1;
	}
}

class App extends React.Component{

	constructor(props){
		super(props);
		this.state = {
			regionCode: 0,
			itemInputs: {
				name: "",
				quantity: 1,
				price: 9.99,
			},
			itemList: [],
			itemId:0,
			total:0,
			promoCode:"",
			discount:0
		}
		this.onInputChange = this.onInputChange.bind(this);
		this.addItem = this.addItem.bind(this);
		this.addRandom = this.addRandom.bind(this);
		this.removeItem = this.removeItem.bind(this);
		this.addPromo = this.addPromo.bind(this);
		this.changeRegion = this.changeRegion.bind(this);
		this.changeTotal = this.changeTotal.bind(this);
		this.checkout = this.checkout.bind(this);
	}
	
	onInputChange(name, value, type){
		if (type===1){	
		    const { itemInputs } = { ...this.state };
		    const currentState = itemInputs;
		    currentState[name] = value;
		    this.setState({ itemInputs: currentState });
		}else{
			this.setState({[name]:value})
		}
	}

	changeTotal(amount){
	 	getSubtotal(this.state.total, amount).then((newTot)=>{
	 		if (newTot){
	 			const newT = newTot.newsub
	 			if (newT>=0){
	 				let newTotal = parseFloat(newT).toFixed(4)
	 				this.setState({total:newTotal});
	 			}
	 			
	 		}
	 	});
	}

	addItem(name, quantity, price){
	 	// add to the array
	 	let id = this.state.itemId;
	 	const newItem = new Item(name, parseInt(quantity, 10), parseFloat(price).toFixed(2), id);
	 	let items = this.state.itemList.push(newItem);
	 	id = id + 1;
	 	this.setState({itemLists:items});
	 	this.setState({itemId:id});
	 	this.changeTotal(price);

	}

	addRandom(addItem){
		const liquorType = ["Whiskey", "Brandy", "Beer", "Gin", "Rum", 
							"Vodka", "Tequila"]

		const liquor = Math.floor(Math.random()*liquorType.length);
		const price = Math.ceil(Math.random()*(liquor+1)*10);
		const quantity = Math.ceil(Math.random()*10);

		addItem(liquorType[liquor], quantity, price);
		alert(quantity + " bottles of " + liquorType[liquor] + " for $" + price+
			" has been added to your cart");
		this.changeTotal(price);
	}

	removeItem(id){
		let itemList = this.state.itemList;
		let price = 0;
		var i=0;
		for (;i<itemList.length;i++){
			if (itemList[i].id === id){//found
				price = itemList[i].price;
				itemList.splice(i, 1);
				break;
			}
		}
		for  (;i<itemList.length;i++){
			itemList[i].decreaseId();
		}
		this.setState({itemList:itemList});
		this.changeTotal(-price);

	}

	addPromo(code){
		const codes = {'20discount': 20, '30discount': 30};
		if (code in codes){
			const discount = codes[code]
			this.setState({discount:discount})
			alert(codes[code] + "% off will be added to your total!");
		}else{
			alert("This promo code does not exist")
		}
		

	}

	changeRegion(regionCode){
		if (regionCode !== this.state.regionCode){
			this.setState({regionCode:regionCode});
		}
	}

	checkout(){
		if (!this.state.total){
			alert("Your Cart is empty!")
		}else if (!this.state.regionCode){
			alert("Please choose your region")
		}else{
			checkout(this.state.total, this.state.discount, this.state.regionCode)
				.then((re)=>{
					if (re){
						const total = re.total.toFixed(2);
						alert("Your total is "+ total);
					}
				})

		}
	}

	render(){
		return (
		    <div className="App">
		      	<Navbar regionCode={this.state.regionCode}
		      			changeRegion={this.changeRegion}
		      			addPromo={this.addPromo} 
		      			promoCode={this.state.promoCode}
		      			onInputChange={this.onInputChange}/>
		      	<AppContainer itemInputs={this.state.itemInputs}
		      				  itemList={this.state.itemList}
		      				  onInputChange={this.onInputChange}
		      				  addItem={this.addItem}
		      				  addRandom={this.addRandom}
		      				  removeItem={this.removeItem}
		      				  total={this.state.total}
		      				  checkout={this.checkout}
		      				  />
		    </div>
		);
	}
  
}

export default App;
