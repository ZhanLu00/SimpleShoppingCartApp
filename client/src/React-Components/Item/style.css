.itemRow {
	height: 30px;
}