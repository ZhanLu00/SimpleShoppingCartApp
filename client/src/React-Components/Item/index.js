import React from 'react';
import './style.css';
import { Row, Col, Button, Card } from 'react-bootstrap';

class Item extends React.Component{

	constructor(props){
		super(props)
		this.state={
			itemId: this.props.itemId
		}
		this.removeItem = this.removeItem.bind(this);
	}

	removeItem(){
		this.props.removeItem(this.state.itemId)
	}

	render(){
		const { itemName, itemQuantity, itemPrice } = this.props;

		return(
			<Card border="info">
				<Row className="itemRow">
					<Col>{itemName}
					</Col>
					<Col>
						Quantity: {itemQuantity}
					</Col>
					<Col>${itemPrice}
					</Col>
					<Col>
						<Button variant="outline-danger" 
								size="sm"
								onClick={this.removeItem}>
							Delete
						</Button>
					</Col>
				</Row>
			</Card>
		)
	}
}

export default Item;