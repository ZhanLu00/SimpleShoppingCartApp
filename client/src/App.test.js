import React from 'react';
import { render } from '@testing-library/react';
import App from './App';
import Item from './React-Components/Item'
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';

test('components renders properly', () => {
	const app = renderer.create(
	  <App />,
	);
	let tree = app.toJSON();
	expect(tree).toMatchSnapshot();

  	const item = renderer.create(
		<Item itemName="Chips" itemQuantity={1} itemPrice={2} itemId={1}>
		</Item>,
  	);
  	tree = item.toJSON();
  	expect(tree).toMatchSnapshot();

});

it('Checking itemlists after adding 2 items ', () => {
	 const wrapper = renderer.create(<App />);
	 const testInstance = wrapper.getInstance();

	let item1 = { name: "chips",
				quantity: 1,
				price: 9.99};
	let item2 = { name: "fried chicken",
				quantity: 2,
				price: 20.00};

	testInstance.setState({ itemInputs: item1 });
	testInstance.addItem(item1.name, item1.quantity, item1.price);
	
	testInstance.setState({ itemInputs: item2 });
	testInstance.addItem(item2.name, item2.quantity, item2.price);
	
	expect(testInstance.state.itemList.length).toBe(2);

});


