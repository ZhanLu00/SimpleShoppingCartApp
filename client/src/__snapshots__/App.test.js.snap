// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components renders properly 1`] = `
<div
  className="App"
>
  <div>
    <nav
      className="navbar navbar-expand-sm navbar-dark bg-dark sticky-top"
    >
      <span
        className="navbarBrand navbar-brand"
        href=""
      >
        <img
          alt=""
          className="d-inline-block align-top"
          src="https://cdn.clipart.email/d8a63873616061abe7766d804101fd26_online-shopping-cart-logo-png-png-image_256-256.png"
        />
         
        Checkout App
      </span>
      <button
        aria-controls="basic-navbar-nav"
        aria-label="Toggle navigation"
        className="navbar-toggler collapsed"
        onClick={[Function]}
        type="button"
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        aria-expanded={null}
        className="navbar-collapse collapse"
        id="basic-navbar-nav"
      >
        <div
          className="mr-auto navbar-nav"
          onKeyDown={[Function]}
        >
          <a
            className="nav-link"
            data-rb-event-key={null}
            href="#"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
          >
            Home
          </a>
          <div
            className="dropdown nav-item"
            onKeyDown={[Function]}
          >
            <a
              aria-expanded={false}
              aria-haspopup={true}
              className="dropdown-toggle nav-link"
              data-rb-event-key={null}
              href="#"
              id="basic-nav-dropdown"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="button"
            >
              Select Region
            </a>
          </div>
        </div>
        <form
          className="form-inline"
        >
          <div
            className="nav-discount"
          >
            <input
              className="mr-sm-2 nav-discount-input form-control form-control-sm"
              id="promoCode"
              onChange={[Function]}
              placeholder="promo code"
              required={true}
              type="text"
              value=""
            />
            <button
              className="nav-discount-submit btn btn-outline-success btn-sm"
              disabled={false}
              onClick={[Function]}
              type="submit"
            >
              Discount
            </button>
          </div>
        </form>
      </div>
    </nav>
  </div>
  <div
    className="appContainer container"
  >
    <div
      className="row"
    >
      <div
        className="newItemContainer col-xl-5 col-lg-6 col-md-4 col-sm-12"
      >
        <div
          className="newItemContainer"
        >
          <form
            className="newItemForm"
          >
            <div
              className="form-group"
            >
              <label
                className="form-label"
              >
                Item Name
              </label>
              <input
                className="form-control"
                id="name"
                onChange={[Function]}
                placeholder="Enter Item Name"
                required={true}
                type="text"
                value=""
              />
            </div>
            <div
              className="form-row"
            >
              <div
                className="col-lg-6 col-md-12 col-sm-6"
              >
                <div
                  className="form-group"
                >
                  <label
                    className="form-label"
                  >
                    Item Quantity
                  </label>
                  <input
                    className="form-control"
                    id="quantity"
                    min={1}
                    onChange={[Function]}
                    required={true}
                    type="number"
                    value={1}
                  />
                </div>
              </div>
              <div
                className="col-lg-6 col-md-12 col-sm-6"
              >
                <div
                  className="form-group"
                >
                  <label
                    className="form-label"
                  >
                    Item Price
                  </label>
                  <input
                    className="form-control"
                    id="price"
                    min={0.01}
                    onChange={[Function]}
                    required={true}
                    step={0.05}
                    type="number"
                    value={9.99}
                  />
                </div>
              </div>
            </div>
            <button
              className="btn btn-primary"
              disabled={false}
              onClick={[Function]}
              type="submit"
            >
              Add to your chart
            </button>
            <small
              className="form-text"
            >
              Or
            </small>
            <button
              className="btn btn-primary"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Add Random
            </button>
          </form>
        </div>
      </div>
      <div
        className="checkoutContainer col-xl-7 col-lg-6 col-md-8 col-sm-12"
      >
        <div
          className="itemContainer"
        >
          <nav
            className="itemNav navbar navbar-expand navbar-dark bg-info"
          >
            <span
              className="navbar-brand"
            >
              Your Cart
            </span>
          </nav>
          <div
            className="itemList"
          />
          <nav
            className="itemTotal navbar navbar-expand navbar-dark bg-danger sticky-bottom fixed-buttom"
          >
            <span
              className="navbar-brand"
            >
              Your Subtotal: $
              0.00
            </span>
            <div
              aria-expanded={null}
              className="justify-content-end navbar-collapse collapse"
            >
              <button
                className="btn btn-light"
                disabled={false}
                onClick={[Function]}
                type="submit"
              >
                Checkout
              </button>
            </div>
          </nav>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components renders properly 2`] = `
<div
  className="card border-info"
>
  <div
    className="itemRow row"
  >
    <div
      className="col"
    >
      Chips
    </div>
    <div
      className="col"
    >
      Quantity: 
      1
    </div>
    <div
      className="col"
    >
      $
      2
    </div>
    <div
      className="col"
    >
      <button
        className="btn btn-outline-danger btn-sm"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
